Mathlib/FieldTheory/SeparableDegree.lean

```diff
@@ -1,147 +1,404 @@
+-- TODO: `IsAlgClosed K` should be replaced by `∀ α : E, Splits (algebraMap F K) (minpoly F α)`
+def embEquivOfIsAlgClosed (halg : Algebra.IsAlgebraic F E)
+    (K : Type w) [Field K] [Algebra F K] [IsAlgClosed K] :
+    Emb F E ≃ (E →ₐ[F] K) := by
+  have hac := IsAlgClosure.ofAlgebraic F E (AlgebraicClosure E) halg
+  let i := IsAlgClosed.lift (M := K) hac.algebraic
+  have h2 : Function.Injective (i.comp (A := E)) := fun j j' h ↦ by
+    apply_fun fun x ↦ (x : E → K) at h
+    simp only [AlgHom.coe_comp] at h
+    exact AlgHom.ext (congrFun (Function.Injective.comp_left i.injective h))
+  let i' := AlgEquiv.ofInjectiveField i
+  let toFun : (E →ₐ[F] (AlgebraicClosure E)) → (E →ₐ[F] K) :=
+    fun f ↦ i.comp f
+  let invFun : (E →ₐ[F] K) → (E →ₐ[F] (AlgebraicClosure E)) :=
+    fun f ↦ i'.symm.toAlgHom.comp (IntermediateField.inclusion <|
+      test5 halg f i) |>.comp f.rangeRestrict
+  have right_inv : Function.RightInverse invFun toFun := fun f ↦ by
+    simp only
+    rw [AlgHom.comp_assoc, ← AlgHom.comp_assoc, test1]
+    rfl
+  exact ⟨toFun, invFun, fun f ↦ h2 <| right_inv <| toFun f, right_inv⟩
+
+/-- The separable degree of `E / F` is equal to the cardinality of `E →ₐ[F] K` when `E / F`
+is algebraic and `K / F` is algebraically closed. -/
+theorem sepDegree_eq_of_isAlgClosed (halg : Algebra.IsAlgebraic F E)
+    (K : Type w) [Field K] [Algebra F K] [IsAlgClosed K] :
+    Cardinal.lift.{w} (sepDegree F E) = Cardinal.mk (E →ₐ[F] K) := by
+  simpa only [Cardinal.mk_uLift] using ((Equiv.ulift.{w} (α := Emb F E)).trans
+    (embEquivOfIsAlgClosed F E halg K)).cardinal_eq
+
+/-- The `finSepDegree F E` is equal to the cardinality of `E →ₐ[F] K` as a natural number,
+when `E / F` is algebraic and `K / F` is algebraically closed. -/
+theorem finSepDegree_eq_of_isAlgClosed (halg : Algebra.IsAlgebraic F E)
+    (K : Type w) [Field K] [Algebra F K] [IsAlgClosed K] :
+    finSepDegree F E = Cardinal.toNat (Cardinal.mk (E →ₐ[F] K)) := by
+  simpa only [Cardinal.toNat_lift] using congr_arg Cardinal.toNat
+    (sepDegree_eq_of_isAlgClosed F E halg K)
+
+def algHomEquivNormalClosure (K : Type w) [Field K] [Algebra F K] :
+    (E →ₐ[F] K) ≃ (E →ₐ[F] (normalClosure F E K)) :=
+  ⟨fun f ↦ (IntermediateField.inclusion f.fieldRange_le_normalClosure) |>.comp
+    (AlgEquiv.ofInjectiveField f).toAlgHom, normalClosure F E K |>.val.comp,
+      fun _ ↦ rfl, fun _ ↦ rfl⟩
+
+lemma test101 (K : Type w) [Field K] [Algebra F K]
+    {S : Set E} (hS : adjoin F S = ⊤) :
+    adjoin F (⨆ f : E →ₐ[F] K, f '' S) = normalClosure F E K := by
+  simp only [normalClosure, gc.l_iSup]
+  congr
+  funext f
+  rw [← adjoin_map, hS, AlgHom.fieldRange_eq_map]
+
+lemma test102 (K : Type w) [Field K] [Algebra F K] {S : Set E} (hS : adjoin F S = ⊤)
+    (hK : ∀ s ∈ S, IsIntegral F s ∧ Polynomial.Splits (algebraMap F K) (minpoly F s)) :
+    ∀ s ∈ S, Polynomial.Splits (algebraMap F (normalClosure F E K)) (minpoly F s) := by
+  intro s hs
+  sorry
```

Comment:
I don't know how to prove this result in a simple way. I believe it's correct, but my plan is divide it into 3 steps:

- First is to prove the case that `K = AlgebraicClosure F` and `E` is an intermediate field of these two fields. Then I can use the fact that `K / F` is normal.
- Second step is `K / F` algebraic case. In this case one can show that the normal closure of `E` in `AlgebraicClosure F` is in fact equal to the normal closure of `E` in `K`.
- Third step is for general case, in this case consider the maximal algebraic subextension of `K / F` (do we have this in mathlib yet?)

Reply:
I'm planning to use this [refined version](https://github.com/leanprover-community/mathlib4/pull/8221/commits/dbe017bf8de59e5bcff758aec2a988f202e339fd#diff-155a98da5b1e87cf16b3ed92d347e59836382af78331f95e267d330a7b85f782R1145) of `algHom_mk_adjoin_splits` to prove [this](https://github.com/leanprover-community/mathlib4/compare/IsNormalClosure?expand=1#diff-d1090b16dffcfbd351531a9ffbd5909e1a2ec8075f6893ee523df3a9d5326dd3R56-R57) which implies the roots all lie in the normalClosure.

Re: your 3rd step, you can use [integralClosure](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/IntegralClosure.html#integralClosure). (Actually I don't know whether the algebraic elements over a CommRing always form a ring; maybe you can show that using resolvents?)

Reply:
> Actually I don't know whether the algebraic elements over a CommRing always form a ring; maybe you can show that using resolvents?

At least for field case it's true. For example the `Qbar` in `C` is a field. The proof is quite standard, and I think it must be already in mathlib.

Reply:
[integralClosure](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/IntegralClosure.html#integralClosure) by definition consists of exactly the integral elements, which are the same as the algebraic elements over a field, right?

Reply:
We're apparently lacking a constructor that takes a Subalgebra in a field extension and an assumption that it's algebraic (equivalently, each element is algebraic in the ambient field), and produces an IntermediateField with the same carrier. I think it could be used to golf lots of stuff, such as [IntermediateField.sup_toSubalgebra](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Adjoin.html#IntermediateField.sup_toSubalgebra).

Reply:
I use your `IntermediateField.algHom_mk_adjoin_splits₀'` and the proof is almost finished. But there are still some gaps:

- I need `IntermediateField.adjoin.liftSingleton` but there is only `Algebra.adjoin.liftSingleton`
- ... and that although `Algebra.adjoin.liftSingleton` is provided, but there is no any lemma says that `Algebra.adjoin.liftSingleton F s a` maps `s` to `a` ?!?

Reply:
Thanks! I didn't got much time to work on IsNormalClosure and you've overtaken me on the development of normalClosure API. Most of your results should be moved to the normalClosure file though. When I open the PR I'll do that and add you as a co-author.

> I need `IntermediateField.adjoin.liftSingleton` but there is only `Algebra.adjoin.liftSingleton`

Actually there is [IntermediateField.algHomAdjoinIntegralEquiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Adjoin.html#IntermediateField.algHomAdjoinIntegralEquiv), which is what I use in the new convenience lemma [algHom_mk_adjoin_splits_of_aeval](https://github.com/leanprover-community/mathlib4/pull/8221/files#diff-155a98da5b1e87cf16b3ed92d347e59836382af78331f95e267d330a7b85f782R1162-R1168) I added in #8221. It should be easier to use `algHom_mk_of_splits_of_aeval` there.

- ... and that although `Algebra.adjoin.liftSingleton` is provided, but there is no any lemma says that `Algebra.adjoin.liftSingleton F s a` maps `s` to `a` ?!?

Yeah that seems to be a one-off construction and only used in `maximalSubfieldWithHom_eq_top` which is to be removed.

The current APIs can be incomplete at places (`algHomAdjoinIntegralEquiv_symm_apply_gen` was added by me recently), but this will no longer be the case once we address #8075. In the short term, we may add IsAdjoinRoot "instances" for Algebra.adjoin (assuming integrality) and IntermediateField.adjoin to be able to use the IsAdjoinRoot API.

Reply:
> Actually there is [IntermediateField.algHomAdjoinIntegralEquiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Adjoin.html#IntermediateField.algHomAdjoinIntegralEquiv) ... (`algHomAdjoinIntegralEquiv_symm_apply_gen` was added by me recently) ...

Thanks, I use these two to finally make the proof sorry-free.

> Thanks! I didn't get much time to work on IsNormalClosure since the week started and you've overtaken me on the development of normalClosure API. Most of your results should be moved to the normalClosure file though. When I open the PR I'll do that and add you as a co-author.
 
You're welcome. Feel free to take the relevant part to the normal closure file. After that I'll make corresponding changes in this file. :)